[BUG]: stl_bind generates invalid type signature for Sphinx/Pylance

### Required prerequisites

- [X] Make sure you've read the [documentation](https://pybind11.readthedocs.io). Your issue may be addressed there.
- [X] Search the [issue tracker](https://github.com/pybind/pybind11/issues) and [Discussions](https:/pybind/pybind11/discussions) to verify that this hasn't already been reported. +1 or comment there if it has.
- [X] Consider asking first in the [Gitter chat room](https://gitter.im/pybind/Lobby) or in a [Discussion](https:/pybind/pybind11/discussions/new).

### Problem description

STL_bind creates subclasses that try to mimic keyview. However, our subclasses have no explicit relation with typing.collections.KeyView or collection.abc.KeyView etc... https://github.com/pybind/pybind11/pull/3985 attempts to fix it by removing the [], but this seems like a cludge and we really should let the typing system represent these as the correct type. I am wondering if there is a way we can still have Sphinx / Pylance recognize these KeyView, ItemViews etc as what they are instead of just an opaque class. Maybe just having them inherit from / mixin collections.abc.KeyViews would be appropiate?

Would appreciate thoughts on this front @rwgk @henryiii?

### Reproducible example code

_No response_

I found a workaround to the linker error, but if I fix the typing system, we get an error because we define the same classname multiple times:

diff:
```
diff --git a/include/pybind11/stl_bind.h b/include/pybind11/stl_bind.h
index 22a29b47..bc1e0312 100644
--- a/include/pybind11/stl_bind.h
+++ b/include/pybind11/stl_bind.h
@@ -651,6 +651,15 @@ struct items_view {
     Map &map;
 };

+/*
+// Helper function needed to prevent linker error
+template <typename Value>
+const char* get_caster_name(){
+       static constexpr auto value_name = detail::make_caster<Value>::name;
+       return value_name.text;
+}
+*/
+
 PYBIND11_NAMESPACE_END(detail)

 template <typename Map, typename holder_type = std::unique_ptr<Map>, typename... Args>
@@ -662,6 +671,12 @@ class_<Map, holder_type> bind_map(handle scope, const std::string &name, Args &&
     using ItemsView = detail::items_view<Map>;
     using Class_ = class_<Map, holder_type>;

+       // Need to do an explicit copy from static to local to prevent linker errors
+       static constexpr auto key_conv_name = detail::make_caster<KeyType>::name;
+       static constexpr auto value_conv_name = detail::make_caster<MappedType>::name;
+       const char *key_name = key_conv_name.text;
+       const char *value_name = value_conv_name.text;
+
     // If either type is a non-module-local bound type then make the map binding non-local as well;
     // otherwise (e.g. both types are either module-local or converting) the map will be
     // module-local.
@@ -674,11 +689,11 @@ class_<Map, holder_type> bind_map(handle scope, const std::string &name, Args &&

     Class_ cl(scope, name.c_str(), pybind11::module_local(local), std::forward<Args>(args)...);
     class_<KeysView> keys_view(
-        scope, ("KeysView[" + name + "]").c_str(), pybind11::module_local(local));
+        scope, (std::string("KeysView[") + key_name + ']').c_str(), pybind11::module_local(local));
     class_<ValuesView> values_view(
-        scope, ("ValuesView[" + name + "]").c_str(), pybind11::module_local(local));
+        scope, (std::string("ValuesView[") + value_name + ']').c_str(), pybind11::module_local(local));
     class_<ItemsView> items_view(
-        scope, ("ItemsView[" + name + "]").c_str(), pybind11::module_local(local));
+        scope, (std::string("ItemsView[") + key_name + ',' + value_name + ']').c_str(), pybind11::module_local(local));

     cl.def(init<>());
```
Error:
```
    import pybind11_tests
E   ImportError: generic_type: cannot initialize type "KeysView[str]": an object with that name is already defined
```

--- a/include/pybind11/detail/type_caster_base.h
+++ b/include/pybind11/detail/type_caster_base.h
@@ -1011,9 +1011,8 @@ PYBIND11_NOINLINE std::string type_info_description(const std::type_info &ti) {
         handle th((PyObject *) type_data->type);
         return th.attr("__module__").cast<std::string>() + '.'
                + th.attr("__qualname__").cast<std::string>();
-    } else {
-        return clean_type_id(ti.name());
     }
+    return clean_type_id(ti.name());
 }
 
 PYBIND11_NAMESPACE_END(detail)

--- a/include/pybind11/detail/type_caster_base.h
+++ b/include/pybind11/detail/type_caster_base.h
@@ -1006,5 +1006,14 @@ protected:
     static Constructor make_move_constructor(...) { return nullptr; }
 };
 
+PYBIND11_NOINLINE std::string type_info_description(const std::type_info& ti) {
+    if (auto *type_data = get_type_info(ti)) {
+        handle th((PyObject *) type_data->type);
+        return th.attr("__module__").cast<std::string>() + '.' + th.attr("__qualname__").cast<std::string>();
+    } else {
+        return clean_type_id(ti.name());
+    }
+}
+
 PYBIND11_NAMESPACE_END(detail)
 PYBIND11_NAMESPACE_END(PYBIND11_NAMESPACE)

--- a/include/pybind11/detail/type_caster_base.h
+++ b/include/pybind11/detail/type_caster_base.h
@@ -1006,10 +1006,11 @@ protected:
     static Constructor make_move_constructor(...) { return nullptr; }
 };
 
-PYBIND11_NOINLINE std::string type_info_description(const std::type_info& ti) {
+PYBIND11_NOINLINE std::string type_info_description(const std::type_info &ti) {
     if (auto *type_data = get_type_info(ti)) {
         handle th((PyObject *) type_data->type);
-        return th.attr("__module__").cast<std::string>() + '.' + th.attr("__qualname__").cast<std::string>();
+        return th.attr("__module__").cast<std::string>() + '.'
+               + th.attr("__qualname__").cast<std::string>();
     } else {
         return clean_type_id(ti.name());
     }

--- a/include/pybind11/stl_bind.h
+++ b/include/pybind11/stl_bind.h
@@ -727,7 +727,7 @@ class_<Map, holder_type> bind_map(handle scope, const std::string &name, Args &&
     // If key type isn't properly wrapped, fall back to C++ names
     if (key_type_name == "%") {
         if (auto *key_info = detail::get_type_info(typeid(KeyType))) {
-            handle th((PyObject *) tinfo->type);
+            handle th((PyObject *) key_info->type);
             key_type_name = th.attr("__module__").cast<std::string>() + "."
                             + th.attr("__qualname__").cast<std::string>();
         } else {
@@ -737,7 +737,7 @@ class_<Map, holder_type> bind_map(handle scope, const std::string &name, Args &&
     // Similarly for value type:
     if (mapped_type_name == "%") {
         if (auto *mapped_info = detail::get_type_info(typeid(MappedType))) {
-            handle th((PyObject *) tinfo->type);
+            handle th((PyObject *) mapped_info->type);
             mapped_type_name = th.attr("__module__").cast<std::string>() + "."
                                + th.attr("__qualname__").cast<std::string>();
         } else {

--- a/include/pybind11/stl_bind.h
+++ b/include/pybind11/stl_bind.h
@@ -10,7 +10,7 @@
 #pragma once
 
 #include "detail/common.h"
-#include "detail/typeid.h"
+#include "detail/type_caster_base.h"
 #include "cast.h"
 #include "operators.h"
 
@@ -724,23 +724,11 @@ class_<Map, holder_type> bind_map(handle scope, const std::string &name, Args &&
 
     // If key type isn't properly wrapped, fall back to C++ names
     if (key_type_name == "%") {
-        if (auto *key_info = detail::get_type_info(typeid(KeyType))) {
-            handle th((PyObject *) key_info->type);
-            key_type_name = th.attr("__module__").cast<std::string>() + '.'
-                            + th.attr("__qualname__").cast<std::string>();
-        } else {
-            key_type_name = type_id<KeyType>();
-        }
+        key_type_name = detail::type_info_description(typeid(KeyType));
     }
     // Similarly for value type:
     if (mapped_type_name == "%") {
-        if (auto *mapped_info = detail::get_type_info(typeid(MappedType))) {
-            handle th((PyObject *) mapped_info->type);
-            mapped_type_name = th.attr("__module__").cast<std::string>() + '.'
-                               + th.attr("__qualname__").cast<std::string>();
-        } else {
-            mapped_type_name = type_id<MappedType>();
-        }
+        mapped_type_name = detail::type_info_description(typeid(MappedType));
     }
 
     // Wrap KeysView[KeyType] if it wasn't already wrapped

--- a/include/pybind11/stl_bind.h
+++ b/include/pybind11/stl_bind.h
@@ -667,13 +667,7 @@ struct KeysViewImpl : public KeysView {
     explicit KeysViewImpl(Map &map) : map(map) {}
     size_t len() override { return map.size(); }
     iterator iter() override { return make_key_iterator(map.begin(), map.end()); }
-    bool contains(const typename Map::key_type &k) override {
-        auto it = map.find(k);
-        if (it == map.end()) {
-            return false;
-        }
-        return true;
-    }
+    bool contains(const typename Map::key_type &k) override { return map.find(k) != map.end(); }
     bool contains(const object &) override { return false; }
     Map &map;
 };

--- a/include/pybind11/stl_bind.h
+++ b/include/pybind11/stl_bind.h
@@ -726,11 +726,23 @@ class_<Map, holder_type> bind_map(handle scope, const std::string &name, Args &&
 
     // If key type isn't properly wrapped, fall back to C++ names
     if (key_type_name == "%") {
-        key_type_name = type_id<KeyType>();
+        if (auto *key_info = detail::get_type_info(typeid(KeyType))) {
+            handle th((PyObject *) tinfo->type);
+            key_type_name = th.attr("__module__").cast<std::string>() + "."
+                            + th.attr("__qualname__").cast<std::string>();
+        } else {
+            key_type_name = type_id<KeyType>();
+        }
     }
     // Similarly for value type:
     if (mapped_type_name == "%") {
-        mapped_type_name = type_id<MappedType>();
+        if (auto *mapped_info = detail::get_type_info(typeid(MappedType))) {
+            handle th((PyObject *) tinfo->type);
+            mapped_type_name = th.attr("__module__").cast<std::string>() + "."
+                               + th.attr("__qualname__").cast<std::string>();
+        } else {
+            mapped_type_name = type_id<MappedType>();
+        }
     }
 
     // Wrap KeysView[KeyType] if it wasn't already wrapped

--- a/include/pybind11/stl_bind.h
+++ b/include/pybind11/stl_bind.h
@@ -668,9 +668,13 @@ template <typename Map, typename holder_type = std::unique_ptr<Map>, typename...
 class_<Map, holder_type> bind_map(handle scope, const std::string &name, Args &&...args) {
     using KeyType = typename Map::key_type;
     using MappedType = typename Map::mapped_type;
-    using KeysView = detail::keys_view<KeyType>;
-    using ValuesView = detail::values_view<MappedType>;
-    using ItemsView = detail::items_view<KeyType, MappedType>;
+    using StrippedKeyType =
+        typename std::remove_cv<typename std::remove_reference<KeyType>::type>::type;
+    using StrippedMappedType =
+        typename std::remove_cv<typename std::remove_reference<MappedType>::type>::type;
+    using KeysView = detail::keys_view<StrippedKeyType>;
+    using ValuesView = detail::values_view<StrippedMappedType>;
+    using ItemsView = detail::items_view<StrippedKeyType, StrippedMappedType>;
     using Class_ = class_<Map, holder_type>;
 
     // If either type is a non-module-local bound type then make the map binding non-local as well;

--- a/include/pybind11/stl_bind.h
+++ b/include/pybind11/stl_bind.h
@@ -662,8 +662,8 @@ struct items_view {
     virtual ~items_view() = default;
 };
 
-template <typename Map>
-struct KeysViewImpl : public keys_view<typename Map::key_type> {
+template <typename Map, typename KeysView>
+struct KeysViewImpl : public KeysView {
     explicit KeysViewImpl(Map &map) : map(map) {}
     size_t __len__() override { return map.size(); }
     iterator __iter__() override { return make_key_iterator(map.begin(), map.end()); }
@@ -678,16 +678,16 @@ struct KeysViewImpl : public keys_view<typename Map::key_type> {
     Map &map;
 };
 
-template <typename Map>
-struct ValuesViewImpl : public values_view<typename Map::mapped_type> {
+template <typename Map, typename ValuesView>
+struct ValuesViewImpl : public ValuesView {
     explicit ValuesViewImpl(Map &map) : map(map) {}
     size_t __len__() override { return map.size(); }
     iterator __iter__() override { return make_value_iterator(map.begin(), map.end()); }
     Map &map;
 };
 
-template <typename Map>
-struct ItemsViewImpl : public items_view<typename Map::key_type, typename Map::mapped_type> {
+template <typename Map, typename ItemsView>
+struct ItemsViewImpl : public ItemsView {
     explicit ItemsViewImpl(Map &map) : map(map) {}
     size_t __len__() override { return map.size(); }
     iterator __iter__() override { return make_iterator(map.begin(), map.end()); }
@@ -763,7 +763,7 @@ class_<Map, holder_type> bind_map(handle scope, const std::string &name, Args &&
     if (!detail::get_type_info(typeid(ItemsView))) {
         class_<ItemsView> items_view(
             scope,
-            ("ItemsView[" + key_type_name + ", " + mapped_type_name + "]").c_str(),
+            ("ItemsView[" + key_type_name + ", ").append(mapped_type_name + "]").c_str(),
             pybind11::module_local(local));
         items_view.def("__len__", &ItemsView::__len__);
         items_view.def("__iter__",
@@ -790,19 +790,25 @@ class_<Map, holder_type> bind_map(handle scope, const std::string &name, Args &&
 
     cl.def(
         "keys",
-        [](Map &m) { return std::unique_ptr<KeysView>(new detail::KeysViewImpl<Map>(m)); },
+        [](Map &m) {
+            return std::unique_ptr<KeysView>(new detail::KeysViewImpl<Map, KeysView>(m));
+        },
         keep_alive<0, 1>() /* Essential: keep map alive while view exists */
     );
 
     cl.def(
         "values",
-        [](Map &m) { return std::unique_ptr<ValuesView>(new detail::ValuesViewImpl<Map>(m)); },
+        [](Map &m) {
+            return std::unique_ptr<ValuesView>(new detail::ValuesViewImpl<Map, ValuesView>(m));
+        },
         keep_alive<0, 1>() /* Essential: keep map alive while view exists */
     );
 
     cl.def(
         "items",
-        [](Map &m) { return std::unique_ptr<ItemsView>(new detail::ItemsViewImpl<Map>(m)); },
+        [](Map &m) {
+            return std::unique_ptr<ItemsView>(new detail::ItemsViewImpl<Map, ItemsView>(m));
+        },
         keep_alive<0, 1>() /* Essential: keep map alive while view exists */
     );
 

--- a/include/pybind11/stl_bind.h
+++ b/include/pybind11/stl_bind.h
@@ -692,6 +692,15 @@ class_<Map, holder_type> bind_map(handle scope, const std::string &name, Args &&
          mapped_type_descr = detail::make_caster<MappedType>::name;
     std::string key_type_name(key_type_descr.text), mapped_type_name(mapped_type_descr.text);
 
+    // If key type isn't properly wrapped, fall back to C++ names
+    if (key_type_name == "%") {
+        key_type_name = type_id<KeyType>();
+    }
+    // Similarly for value type:
+    if (mapped_type_name == "%") {
+        mapped_type_name = type_id<MappedType>();
+    }
+
     // Wrap KeysView[KeyType] if it wasn't already wrapped
     if (!detail::get_type_info(typeid(KeysView))) {
         class_<KeysView> keys_view(

--- a/include/pybind11/stl_bind.h
+++ b/include/pybind11/stl_bind.h
@@ -720,8 +720,8 @@ class_<Map, holder_type> bind_map(handle scope, const std::string &name, Args &&
     }
 
     Class_ cl(scope, name.c_str(), pybind11::module_local(local), std::forward<Args>(args)...);
-    static constexpr auto key_type_descr = detail::make_caster<KeyType>::name;
-    static constexpr auto mapped_type_descr = detail::make_caster<MappedType>::name;
+    auto key_type_descr = detail::make_caster<KeyType>::name;
+    auto mapped_type_descr = detail::make_caster<MappedType>::name;
     std::string key_type_name(key_type_descr.text), mapped_type_name(mapped_type_descr.text);
 
     // If key type isn't properly wrapped, fall back to C++ names

--- a/include/pybind11/stl_bind.h
+++ b/include/pybind11/stl_bind.h
@@ -688,8 +688,8 @@ class_<Map, holder_type> bind_map(handle scope, const std::string &name, Args &&
     }
 
     Class_ cl(scope, name.c_str(), pybind11::module_local(local), std::forward<Args>(args)...);
-    auto key_type_descr = detail::make_caster<KeyType>::name,
-         mapped_type_descr = detail::make_caster<MappedType>::name;
+    auto key_type_descr = detail::make_caster<KeyType>::name;
+    auto mapped_type_descr = detail::make_caster<MappedType>::name;
     std::string key_type_name(key_type_descr.text), mapped_type_name(mapped_type_descr.text);
 
     // If key type isn't properly wrapped, fall back to C++ names

--- a/include/pybind11/stl_bind.h
+++ b/include/pybind11/stl_bind.h
@@ -641,56 +641,56 @@ auto map_if_insertion_operator(Class_ &cl, std::string const &name)
 
 template <typename KeyType>
 struct keys_view {
-    virtual size_t __len__() = 0;
-    virtual iterator __iter__() = 0;
-    virtual bool __contains__(const KeyType &k) = 0;
-    virtual bool __contains__(const object &k) = 0;
+    virtual size_t len() = 0;
+    virtual iterator iter() = 0;
+    virtual bool contains(const KeyType &k) = 0;
+    virtual bool contains(const object &k) = 0;
     virtual ~keys_view() = default;
 };
 
 template <typename MappedType>
 struct values_view {
-    virtual size_t __len__() = 0;
-    virtual iterator __iter__() = 0;
+    virtual size_t len() = 0;
+    virtual iterator iter() = 0;
     virtual ~values_view() = default;
 };
 
 template <typename KeyType, typename MappedType>
 struct items_view {
-    virtual size_t __len__() = 0;
-    virtual iterator __iter__() = 0;
+    virtual size_t len() = 0;
+    virtual iterator iter() = 0;
     virtual ~items_view() = default;
 };
 
 template <typename Map, typename KeysView>
 struct KeysViewImpl : public KeysView {
     explicit KeysViewImpl(Map &map) : map(map) {}
-    size_t __len__() override { return map.size(); }
-    iterator __iter__() override { return make_key_iterator(map.begin(), map.end()); }
-    bool __contains__(const typename Map::key_type &k) override {
+    size_t len() override { return map.size(); }
+    iterator iter() override { return make_key_iterator(map.begin(), map.end()); }
+    bool contains(const typename Map::key_type &k) override {
         auto it = map.find(k);
         if (it == map.end()) {
             return false;
         }
         return true;
     }
-    bool __contains__(const object &) override { return false; }
+    bool contains(const object &) override { return false; }
     Map &map;
 };
 
 template <typename Map, typename ValuesView>
 struct ValuesViewImpl : public ValuesView {
     explicit ValuesViewImpl(Map &map) : map(map) {}
-    size_t __len__() override { return map.size(); }
-    iterator __iter__() override { return make_value_iterator(map.begin(), map.end()); }
+    size_t len() override { return map.size(); }
+    iterator iter() override { return make_value_iterator(map.begin(), map.end()); }
     Map &map;
 };
 
 template <typename Map, typename ItemsView>
 struct ItemsViewImpl : public ItemsView {
     explicit ItemsViewImpl(Map &map) : map(map) {}
-    size_t __len__() override { return map.size(); }
-    iterator __iter__() override { return make_iterator(map.begin(), map.end()); }
+    size_t len() override { return map.size(); }
+    iterator iter() override { return make_iterator(map.begin(), map.end()); }
     Map &map;
 };
 
@@ -747,25 +747,25 @@ class_<Map, holder_type> bind_map(handle scope, const std::string &name, Args &&
     if (!detail::get_type_info(typeid(KeysView))) {
         class_<KeysView> keys_view(
             scope, ("KeysView[" + key_type_name + "]").c_str(), pybind11::module_local(local));
-        keys_view.def("__len__", &KeysView::__len__);
+        keys_view.def("__len__", &KeysView::len);
         keys_view.def("__iter__",
-                      &KeysView::__iter__,
+                      &KeysView::iter,
                       keep_alive<0, 1>() /* Essential: keep view alive while iterator exists */
         );
         keys_view.def("__contains__",
-                      static_cast<bool (KeysView::*)(const KeyType &)>(&KeysView::__contains__));
+                      static_cast<bool (KeysView::*)(const KeyType &)>(&KeysView::contains));
         // Fallback for when the object is not of the key type
         keys_view.def("__contains__",
-                      static_cast<bool (KeysView::*)(const object &)>(&KeysView::__contains__));
+                      static_cast<bool (KeysView::*)(const object &)>(&KeysView::contains));
     }
     // Similarly for ValuesView:
     if (!detail::get_type_info(typeid(ValuesView))) {
         class_<ValuesView> values_view(scope,
                                        ("ValuesView[" + mapped_type_name + "]").c_str(),
                                        pybind11::module_local(local));
-        values_view.def("__len__", &ValuesView::__len__);
+        values_view.def("__len__", &ValuesView::len);
         values_view.def("__iter__",
-                        &ValuesView::__iter__,
+                        &ValuesView::iter,
                         keep_alive<0, 1>() /* Essential: keep view alive while iterator exists */
         );
     }
@@ -775,9 +775,9 @@ class_<Map, holder_type> bind_map(handle scope, const std::string &name, Args &&
             scope,
             ("ItemsView[" + key_type_name + ", ").append(mapped_type_name + "]").c_str(),
             pybind11::module_local(local));
-        items_view.def("__len__", &ItemsView::__len__);
+        items_view.def("__len__", &ItemsView::len);
         items_view.def("__iter__",
-                       &ItemsView::__iter__,
+                       &ItemsView::iter,
                        keep_alive<0, 1>() /* Essential: keep view alive while iterator exists */
         );
     }

--- a/include/pybind11/stl_bind.h
+++ b/include/pybind11/stl_bind.h
@@ -662,6 +662,38 @@ struct items_view {
     virtual ~items_view() = default;
 };
 
+template <typename Map>
+struct KeysViewImpl : public keys_view<typename Map::key_type> {
+    explicit KeysViewImpl(Map &map) : map(map) {}
+    size_t __len__() override { return map.size(); }
+    iterator __iter__() override { return make_key_iterator(map.begin(), map.end()); }
+    bool __contains__(const typename Map::key_type &k) override {
+        auto it = map.find(k);
+        if (it == map.end()) {
+            return false;
+        }
+        return true;
+    }
+    bool __contains__(const object &) override { return false; }
+    Map &map;
+};
+
+template <typename Map>
+struct ValuesViewImpl : public values_view<typename Map::mapped_type> {
+    explicit ValuesViewImpl(Map &map) : map(map) {}
+    size_t __len__() override { return map.size(); }
+    iterator __iter__() override { return make_value_iterator(map.begin(), map.end()); }
+    Map &map;
+};
+
+template <typename Map>
+struct ItemsViewImpl : public items_view<typename Map::key_type, typename Map::mapped_type> {
+    explicit ItemsViewImpl(Map &map) : map(map) {}
+    size_t __len__() override { return map.size(); }
+    iterator __iter__() override { return make_iterator(map.begin(), map.end()); }
+    Map &map;
+};
+
 PYBIND11_NAMESPACE_END(detail)
 
 template <typename Map, typename holder_type = std::unique_ptr<Map>, typename... Args>
@@ -688,8 +720,8 @@ class_<Map, holder_type> bind_map(handle scope, const std::string &name, Args &&
     }
 
     Class_ cl(scope, name.c_str(), pybind11::module_local(local), std::forward<Args>(args)...);
-    auto key_type_descr = detail::make_caster<KeyType>::name;
-    auto mapped_type_descr = detail::make_caster<MappedType>::name;
+    static constexpr auto key_type_descr = detail::make_caster<KeyType>::name;
+    static constexpr auto mapped_type_descr = detail::make_caster<MappedType>::name;
     std::string key_type_name(key_type_descr.text), mapped_type_name(mapped_type_descr.text);
 
     // If key type isn't properly wrapped, fall back to C++ names
@@ -758,53 +790,19 @@ class_<Map, holder_type> bind_map(handle scope, const std::string &name, Args &&
 
     cl.def(
         "keys",
-        [](Map &m) {
-            struct KeysViewImpl : public KeysView {
-                Map &map;
-                explicit KeysViewImpl(Map &map) : map(map) {}
-                size_t __len__() override { return map.size(); }
-                iterator __iter__() override { return make_key_iterator(map.begin(), map.end()); }
-                bool __contains__(const KeyType &k) override {
-                    auto it = map.find(k);
-                    if (it == map.end()) {
-                        return false;
-                    }
-                    return true;
-                }
-                bool __contains__(const object &) override { return false; }
-            };
-            return std::unique_ptr<KeysView>(new KeysViewImpl(m));
-        },
+        [](Map &m) { return std::unique_ptr<KeysView>(new detail::KeysViewImpl<Map>(m)); },
         keep_alive<0, 1>() /* Essential: keep map alive while view exists */
     );
 
     cl.def(
         "values",
-        [](Map &m) {
-            struct ValuesViewImpl : public ValuesView {
-                Map &map;
-                explicit ValuesViewImpl(Map &map) : map(map) {}
-                size_t __len__() override { return map.size(); }
-                iterator __iter__() override {
-                    return make_value_iterator(map.begin(), map.end());
-                }
-            };
-            return std::unique_ptr<ValuesView>(new ValuesViewImpl(m));
-        },
+        [](Map &m) { return std::unique_ptr<ValuesView>(new detail::ValuesViewImpl<Map>(m)); },
         keep_alive<0, 1>() /* Essential: keep map alive while view exists */
     );
 
     cl.def(
         "items",
-        [](Map &m) {
-            struct ItemsViewImpl : public ItemsView {
-                Map &map;
-                explicit ItemsViewImpl(Map &map) : map(map) {}
-                size_t __len__() override { return map.size(); }
-                iterator __iter__() override { return make_iterator(map.begin(), map.end()); }
-            };
-            return std::unique_ptr<ItemsView>(new ItemsViewImpl(m));
-        },
+        [](Map &m) { return std::unique_ptr<ItemsView>(new detail::ItemsViewImpl<Map>(m)); },
         keep_alive<0, 1>() /* Essential: keep map alive while view exists */
     );
 

--- a/include/pybind11/stl_bind.h
+++ b/include/pybind11/stl_bind.h
@@ -700,10 +700,8 @@ template <typename Map, typename holder_type = std::unique_ptr<Map>, typename...
 class_<Map, holder_type> bind_map(handle scope, const std::string &name, Args &&...args) {
     using KeyType = typename Map::key_type;
     using MappedType = typename Map::mapped_type;
-    using StrippedKeyType =
-        typename std::remove_cv<typename std::remove_reference<KeyType>::type>::type;
-    using StrippedMappedType =
-        typename std::remove_cv<typename std::remove_reference<MappedType>::type>::type;
+    using StrippedKeyType = detail::remove_cvref_t<KeyType>;
+    using StrippedMappedType = detail::remove_cvref_t<MappedType>;
     using KeysView = detail::keys_view<StrippedKeyType>;
     using ValuesView = detail::values_view<StrippedMappedType>;
     using ItemsView = detail::items_view<StrippedKeyType, StrippedMappedType>;

--- a/include/pybind11/stl_bind.h
+++ b/include/pybind11/stl_bind.h
@@ -10,10 +10,13 @@
 #pragma once
 
 #include "detail/common.h"
+#include "detail/typeid.h"
+#include "cast.h"
 #include "operators.h"
 
 #include <algorithm>
 #include <sstream>
+#include <type_traits>
 
 PYBIND11_NAMESPACE_BEGIN(PYBIND11_NAMESPACE)
 PYBIND11_NAMESPACE_BEGIN(detail)
@@ -636,19 +639,27 @@ auto map_if_insertion_operator(Class_ &cl, std::string const &name)
         "Return the canonical string representation of this map.");
 }
 
-template <typename Map>
+template <typename KeyType>
 struct keys_view {
-    Map &map;
+    virtual size_t __len__() = 0;
+    virtual iterator __iter__() = 0;
+    virtual bool __contains__(const KeyType &k) = 0;
+    virtual bool __contains__(const object &k) = 0;
+    virtual ~keys_view() = default;
 };
 
-template <typename Map>
+template <typename MappedType>
 struct values_view {
-    Map &map;
+    virtual size_t __len__() = 0;
+    virtual iterator __iter__() = 0;
+    virtual ~values_view() = default;
 };
 
-template <typename Map>
+template <typename KeyType, typename MappedType>
 struct items_view {
-    Map &map;
+    virtual size_t __len__() = 0;
+    virtual iterator __iter__() = 0;
+    virtual ~items_view() = default;
 };
 
 PYBIND11_NAMESPACE_END(detail)
@@ -657,9 +668,9 @@ template <typename Map, typename holder_type = std::unique_ptr<Map>, typename...
 class_<Map, holder_type> bind_map(handle scope, const std::string &name, Args &&...args) {
     using KeyType = typename Map::key_type;
     using MappedType = typename Map::mapped_type;
-    using KeysView = detail::keys_view<Map>;
-    using ValuesView = detail::values_view<Map>;
-    using ItemsView = detail::items_view<Map>;
+    using KeysView = detail::keys_view<KeyType>;
+    using ValuesView = detail::values_view<MappedType>;
+    using ItemsView = detail::items_view<KeyType, MappedType>;
     using Class_ = class_<Map, holder_type>;
 
     // If either type is a non-module-local bound type then make the map binding non-local as well;
@@ -673,12 +684,48 @@ class_<Map, holder_type> bind_map(handle scope, const std::string &name, Args &&
     }
 
     Class_ cl(scope, name.c_str(), pybind11::module_local(local), std::forward<Args>(args)...);
-    class_<KeysView> keys_view(
-        scope, ("KeysView[" + name + "]").c_str(), pybind11::module_local(local));
-    class_<ValuesView> values_view(
-        scope, ("ValuesView[" + name + "]").c_str(), pybind11::module_local(local));
-    class_<ItemsView> items_view(
-        scope, ("ItemsView[" + name + "]").c_str(), pybind11::module_local(local));
+    auto key_type_descr = detail::make_caster<KeyType>::name,
+         mapped_type_descr = detail::make_caster<MappedType>::name;
+    std::string key_type_name(key_type_descr.text), mapped_type_name(mapped_type_descr.text);
+
+    // Wrap KeysView[KeyType] if it wasn't already wrapped
+    if (!detail::get_type_info(typeid(KeysView))) {
+        class_<KeysView> keys_view(
+            scope, ("KeysView[" + key_type_name + "]").c_str(), pybind11::module_local(local));
+        keys_view.def("__len__", &KeysView::__len__);
+        keys_view.def("__iter__",
+                      &KeysView::__iter__,
+                      keep_alive<0, 1>() /* Essential: keep view alive while iterator exists */
+        );
+        keys_view.def("__contains__",
+                      static_cast<bool (KeysView::*)(const KeyType &)>(&KeysView::__contains__));
+        // Fallback for when the object is not of the key type
+        keys_view.def("__contains__",
+                      static_cast<bool (KeysView::*)(const object &)>(&KeysView::__contains__));
+    }
+    // Similarly for ValuesView:
+    if (!detail::get_type_info(typeid(ValuesView))) {
+        class_<ValuesView> values_view(scope,
+                                       ("ValuesView[" + mapped_type_name + "]").c_str(),
+                                       pybind11::module_local(local));
+        values_view.def("__len__", &ValuesView::__len__);
+        values_view.def("__iter__",
+                        &ValuesView::__iter__,
+                        keep_alive<0, 1>() /* Essential: keep view alive while iterator exists */
+        );
+    }
+    // Similarly for ItemsView:
+    if (!detail::get_type_info(typeid(ItemsView))) {
+        class_<ItemsView> items_view(
+            scope,
+            ("ItemsView[" + key_type_name + ", " + mapped_type_name + "]").c_str(),
+            pybind11::module_local(local));
+        items_view.def("__len__", &ItemsView::__len__);
+        items_view.def("__iter__",
+                       &ItemsView::__iter__,
+                       keep_alive<0, 1>() /* Essential: keep view alive while iterator exists */
+        );
+    }
 
     cl.def(init<>());
 
@@ -698,19 +745,51 @@ class_<Map, holder_type> bind_map(handle scope, const std::string &name, Args &&
 
     cl.def(
         "keys",
-        [](Map &m) { return KeysView{m}; },
+        [](Map &m) {
+            struct KeysViewImpl : public KeysView {
+                Map &map;
+                KeysViewImpl(Map &map) : map(map) {}
+                size_t __len__() { return map.size(); }
+                iterator __iter__() { return make_key_iterator(map.begin(), map.end()); }
+                bool __contains__(const KeyType &k) {
+                    auto it = map.find(k);
+                    if (it == map.end()) {
+                        return false;
+                    }
+                    return true;
+                }
+                bool __contains__(const object &) { return false; }
+            };
+            return std::unique_ptr<KeysView>(new KeysViewImpl(m));
+        },
         keep_alive<0, 1>() /* Essential: keep map alive while view exists */
     );
 
     cl.def(
         "values",
-        [](Map &m) { return ValuesView{m}; },
+        [](Map &m) {
+            struct ValuesViewImpl : public ValuesView {
+                Map &map;
+                ValuesViewImpl(Map &map) : map(map) {}
+                size_t __len__() { return map.size(); }
+                iterator __iter__() { return make_value_iterator(map.begin(), map.end()); }
+            };
+            return std::unique_ptr<ValuesView>(new ValuesViewImpl(m));
+        },
         keep_alive<0, 1>() /* Essential: keep map alive while view exists */
     );
 
     cl.def(
         "items",
-        [](Map &m) { return ItemsView{m}; },
+        [](Map &m) {
+            struct ItemsViewImpl : public ItemsView {
+                Map &map;
+                ItemsViewImpl(Map &map) : map(map) {}
+                size_t __len__() { return map.size(); }
+                iterator __iter__() { return make_iterator(map.begin(), map.end()); }
+            };
+            return std::unique_ptr<ItemsView>(new ItemsViewImpl(m));
+        },
         keep_alive<0, 1>() /* Essential: keep map alive while view exists */
     );
 
@@ -748,36 +827,6 @@ class_<Map, holder_type> bind_map(handle scope, const std::string &name, Args &&
     });
 
     cl.def("__len__", &Map::size);
-
-    keys_view.def("__len__", [](KeysView &view) { return view.map.size(); });
-    keys_view.def(
-        "__iter__",
-        [](KeysView &view) { return make_key_iterator(view.map.begin(), view.map.end()); },
-        keep_alive<0, 1>() /* Essential: keep view alive while iterator exists */
-    );
-    keys_view.def("__contains__", [](KeysView &view, const KeyType &k) -> bool {
-        auto it = view.map.find(k);
-        if (it == view.map.end()) {
-            return false;
-        }
-        return true;
-    });
-    // Fallback for when the object is not of the key type
-    keys_view.def("__contains__", [](KeysView &, const object &) -> bool { return false; });
-
-    values_view.def("__len__", [](ValuesView &view) { return view.map.size(); });
-    values_view.def(
-        "__iter__",
-        [](ValuesView &view) { return make_value_iterator(view.map.begin(), view.map.end()); },
-        keep_alive<0, 1>() /* Essential: keep view alive while iterator exists */
-    );
-
-    items_view.def("__len__", [](ItemsView &view) { return view.map.size(); });
-    items_view.def(
-        "__iter__",
-        [](ItemsView &view) { return make_iterator(view.map.begin(), view.map.end()); },
-        keep_alive<0, 1>() /* Essential: keep view alive while iterator exists */
-    );
 
     return cl;
 }

--- a/include/pybind11/stl_bind.h
+++ b/include/pybind11/stl_bind.h
@@ -720,8 +720,8 @@ class_<Map, holder_type> bind_map(handle scope, const std::string &name, Args &&
     }
 
     Class_ cl(scope, name.c_str(), pybind11::module_local(local), std::forward<Args>(args)...);
-    auto key_type_descr = detail::make_caster<KeyType>::name;
-    auto mapped_type_descr = detail::make_caster<MappedType>::name;
+    static constexpr auto key_type_descr = detail::make_caster<KeyType>::name;
+    static constexpr auto mapped_type_descr = detail::make_caster<MappedType>::name;
     std::string key_type_name(key_type_descr.text), mapped_type_name(mapped_type_descr.text);
 
     // If key type isn't properly wrapped, fall back to C++ names

--- a/include/pybind11/stl_bind.h
+++ b/include/pybind11/stl_bind.h
@@ -761,17 +761,17 @@ class_<Map, holder_type> bind_map(handle scope, const std::string &name, Args &&
         [](Map &m) {
             struct KeysViewImpl : public KeysView {
                 Map &map;
-                KeysViewImpl(Map &map) : map(map) {}
-                size_t __len__() { return map.size(); }
-                iterator __iter__() { return make_key_iterator(map.begin(), map.end()); }
-                bool __contains__(const KeyType &k) {
+                explicit KeysViewImpl(Map &map) : map(map) {}
+                size_t __len__() override { return map.size(); }
+                iterator __iter__() override { return make_key_iterator(map.begin(), map.end()); }
+                bool __contains__(const KeyType &k) override {
                     auto it = map.find(k);
                     if (it == map.end()) {
                         return false;
                     }
                     return true;
                 }
-                bool __contains__(const object &) { return false; }
+                bool __contains__(const object &) override { return false; }
             };
             return std::unique_ptr<KeysView>(new KeysViewImpl(m));
         },
@@ -783,9 +783,11 @@ class_<Map, holder_type> bind_map(handle scope, const std::string &name, Args &&
         [](Map &m) {
             struct ValuesViewImpl : public ValuesView {
                 Map &map;
-                ValuesViewImpl(Map &map) : map(map) {}
-                size_t __len__() { return map.size(); }
-                iterator __iter__() { return make_value_iterator(map.begin(), map.end()); }
+                explicit ValuesViewImpl(Map &map) : map(map) {}
+                size_t __len__() override { return map.size(); }
+                iterator __iter__() override {
+                    return make_value_iterator(map.begin(), map.end());
+                }
             };
             return std::unique_ptr<ValuesView>(new ValuesViewImpl(m));
         },
@@ -797,9 +799,9 @@ class_<Map, holder_type> bind_map(handle scope, const std::string &name, Args &&
         [](Map &m) {
             struct ItemsViewImpl : public ItemsView {
                 Map &map;
-                ItemsViewImpl(Map &map) : map(map) {}
-                size_t __len__() { return map.size(); }
-                iterator __iter__() { return make_iterator(map.begin(), map.end()); }
+                explicit ItemsViewImpl(Map &map) : map(map) {}
+                size_t __len__() override { return map.size(); }
+                iterator __iter__() override { return make_iterator(map.begin(), map.end()); }
             };
             return std::unique_ptr<ItemsView>(new ItemsViewImpl(m));
         },

--- a/include/pybind11/stl_bind.h
+++ b/include/pybind11/stl_bind.h
@@ -728,7 +728,7 @@ class_<Map, holder_type> bind_map(handle scope, const std::string &name, Args &&
     if (key_type_name == "%") {
         if (auto *key_info = detail::get_type_info(typeid(KeyType))) {
             handle th((PyObject *) key_info->type);
-            key_type_name = th.attr("__module__").cast<std::string>() + "."
+            key_type_name = th.attr("__module__").cast<std::string>() + '.'
                             + th.attr("__qualname__").cast<std::string>();
         } else {
             key_type_name = type_id<KeyType>();
@@ -738,7 +738,7 @@ class_<Map, holder_type> bind_map(handle scope, const std::string &name, Args &&
     if (mapped_type_name == "%") {
         if (auto *mapped_info = detail::get_type_info(typeid(MappedType))) {
             handle th((PyObject *) mapped_info->type);
-            mapped_type_name = th.attr("__module__").cast<std::string>() + "."
+            mapped_type_name = th.attr("__module__").cast<std::string>() + '.'
                                + th.attr("__qualname__").cast<std::string>();
         } else {
             mapped_type_name = type_id<MappedType>();

--- a/tests/test_stl_binders.py
+++ b/tests/test_stl_binders.py
@@ -317,17 +317,21 @@ def test_map_view_types():
     map_string_double_const = m.MapStringDoubleConst()
     unordered_map_string_double_const = m.UnorderedMapStringDoubleConst()
 
+    assert map_string_double.keys().__class__.__name__ == "KeysView[str]"
+    assert map_string_double.values().__class__.__name__ == "ValuesView[float]"
+    assert map_string_double.items().__class__.__name__ == "ItemsView[str, float]"
+
     keys_type = type(map_string_double.keys())
-    assert isinstance(unordered_map_string_double.keys(), keys_type)
-    assert isinstance(map_string_double_const.keys(), keys_type)
-    assert isinstance(unordered_map_string_double_const.keys(), keys_type)
+    assert type(unordered_map_string_double.keys()) is keys_type
+    assert type(map_string_double_const.keys()) is keys_type
+    assert type(unordered_map_string_double_const.keys()) is keys_type
 
     values_type = type(map_string_double.values())
-    assert isinstance(unordered_map_string_double.values(), values_type)
-    assert isinstance(map_string_double_const.values(), values_type)
-    assert isinstance(unordered_map_string_double_const.values(), values_type)
+    assert type(unordered_map_string_double.values()) is values_type
+    assert type(map_string_double_const.values()) is values_type
+    assert type(unordered_map_string_double_const.values()) is values_type
 
     items_type = type(map_string_double.items())
-    assert isinstance(unordered_map_string_double.items(), items_type)
-    assert isinstance(map_string_double_const.items(), items_type)
-    assert isinstance(unordered_map_string_double_const.items(), items_type)
+    assert type(unordered_map_string_double.items()) is items_type
+    assert type(map_string_double_const.items()) is items_type
+    assert type(unordered_map_string_double_const.items()) is items_type

--- a/tests/test_stl_binders.py
+++ b/tests/test_stl_binders.py
@@ -309,3 +309,25 @@ def test_map_delitem():
     del um["ua"]
     assert sorted(list(um)) == ["ub"]
     assert sorted(list(um.items())) == [("ub", 2.6)]
+
+
+def test_map_view_types():
+    map_string_double = m.MapStringDouble()
+    unordered_map_string_double = m.UnorderedMapStringDouble()
+    map_string_double_const = m.MapStringDoubleConst()
+    unordered_map_string_double_const = m.UnorderedMapStringDoubleConst()
+
+    keys_type = type(map_string_double.keys())
+    assert isinstance(unordered_map_string_double.keys(), keys_type)
+    assert isinstance(map_string_double_const.keys(), keys_type)
+    assert isinstance(unordered_map_string_double_const.keys(), keys_type)
+
+    values_type = type(map_string_double.values())
+    assert isinstance(unordered_map_string_double.values(), values_type)
+    assert isinstance(map_string_double_const.values(), values_type)
+    assert isinstance(unordered_map_string_double_const.values(), values_type)
+
+    items_type = type(map_string_double.items())
+    assert isinstance(unordered_map_string_double.items(), items_type)
+    assert isinstance(map_string_double_const.items(), items_type)
+    assert isinstance(unordered_map_string_double_const.items(), items_type)

